modules/components/src/Aggs/DatesAgg.js

```diff
@@ -0,0 +1,295 @@
+const Arrow = ({ style }) => (
+  <svg
+    style={style}
+    className="DayPickerNavigation_svg__horizontal DayPickerNavigation_svg__horizontal_1"
+    viewBox="0 0 1000 1000"
+  >
+    <path d="M694.4 242.4l249.1 249.1c11 11 11 21 0 32L694.4 772.7c-5 5-10 7-16 7s-11-2-16-7c-11-11-11-21 0-32l210.1-210.1H67.1c-13 0-23-10-23-23s10-23 23-23h805.4L662.4 274.5c-21-21.1 11-53.1 32-32.1z" />
+  </svg>
+);
+
+class DatesAgg extends React.Component {
+  // needs ref...
+
+  render() {
+    const {
+      field = '',
+      displayName = 'Unnamed Field',
+      buckets = [],
+      collapsible = true,
+      handleDateChange = ({ generateNextSQON }) => null,
+      startDateFromSqon = () => null,
+      endDateFromSqon = () => null,
+      datePickerPosition = 'BOTTOM_LEFT',
+    } = this.props;
+
+    const onDatesSet = ({ startDate, endDate }) => {
+      if (startDate && endDate) {
+        handleDateChange({
+          generateNextSQON: sqon =>
+            replaceSQON(
+              {
+                op: 'and',
+                content: [
+                  {
+                    op: '>=',
+                    content: {
+                      field,
+                      value: momentToBucketDate(startDate.startOf('day')),
+                    },
+                  },
+                  {
+                    op: '<=',
+                    content: {
+                      field,
+                      value: momentToBucketDate(endDate.endOf('day')),
+                    },
+                  },
+                ],
+              },
+              sqon,
+            ),
+        });
+      }
+    };
+
+    const getDateFromSqon = dateToGet => sqon =>
+      sqon?.content
+        ?.filter(({ content: { field: sqonField } }) => {
+          return sqonField === field;
+        })
+        ?.find(
+          ({ op }) => op === (dateToGet === START_DATE_INPUT ? '>=' : '<='),
+        )?.content.value;
+
+    const getInitialRange = () => {
+      const bucketWithMoment = buckets.map(({ key_as_string, ...rest }) => ({
+        ...rest,
+        key_as_string,
+        moment: bucketDateToMoment(key_as_string),
+      }));
+      const minBucket = minBy(bucketWithMoment, ({ moment }) =>
+        moment.valueOf(),
+      );
+      const maxBucket = maxBy(bucketWithMoment, ({ moment }) =>
+        moment.valueOf(),
+      );
+
+      const startFromSqon = startDateFromSqon({
+        getDateFromSqon: getDateFromSqon(START_DATE_INPUT),
+      });
+      const endFromSqon = endDateFromSqon({
+        getDateFromSqon: getDateFromSqon(END_DATE_INPUT),
+      });
+      return {
+        startDate:
+          (startFromSqon && bucketDateToMoment(startFromSqon)) ||
+          minBucket?.moment,
+        endDate:
+          (startFromSqon && bucketDateToMoment(endFromSqon)) ||
+          maxBucket?.moment,
+      };
+    };
+
+    const initialRange = getInitialRange();
+
+    return (
+      <AggsWrapper {...{ displayName, collapsible }}>
+        <Component
```

Comment:
Can you refactor some of this logic to be outside the render? it's getting a bit busy.
This state can be moved to the class, and the functions using it can be moved to class methods :)